### PR TITLE
Replace FQCN strings with ::class constants, Fix tests for SF 4

### LINKED ISSUE #357: Runtime error browsing categories with Symfony 3.4
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle              3.28.0 = 3.28.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.8.0  = 3.8.0  Symfony SonataBlockBundle
sonata-project/cache                     2.0.0  = 2.0.0  Cache library
sonata-project/classification-bundle     3.4.0  = 3.4.0  Symfony SonataClassificationBundle
sonata-project/core-bundle               3.7.1  = 3.7.1  Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.3.0  = 2.3.0  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2  = 1.0.2  Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.2.0  = 3.2.0  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.3.0  = 2.3.0  Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.8.0  = 1.8.0  Lightweight Exporter library
sonata-project/user-bundle               4.0.0  = 4.0.0  Symfony SonataUserBundle
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/assetic-bundle     v2.8.2 = v2.8.2 Integrates Assetic into Symfony2
symfony/monolog-bundle     v3.1.2 = v3.1.2 Symfony MonologBundle
symfony/phpunit-bridge     v3.4.1 ~ v4.0.1 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.6.0 = v1.6.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.6.0 = v1.6.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.6.0 = v1.6.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.6.0 = v1.6.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.6.0 = v1.6.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.6.0 = v1.6.0 Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0 = v3.0.0 Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.6.7 ~ v3.1.6 Symfony SwiftmailerBundle
symfony/symfony            v3.4.1 = v3.4.1 The Symfony PHP framework
```

#### PHP version

```
$ php -v
PHP 7.1.12 (cli) (built: Nov 23 2017 04:33:43) ( NTS MSVC14 (Visual C++ 2015) x86 )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.12, Copyright (c) 1999-2017, by Zend Technologies
```

## Subject
Runtime error browsing categories with Symfony 3.4.

## Steps to reproduce
Fresh installation, go to "Categories" admin.

## Expected results
Page is loaded and you see the categories list.

## Actual results
Runtime exception:

> Unable to load the "Symfony\Bridge\Twig\Form\TwigRenderer" runtime.
> Twig_Environment->getRuntime('Symfony\\Bridge\\Twig\\Form\\TwigRenderer')
> in vendor\sonata-project\classification-bundle\Controller\CategoryAdminController.php (line 140)

error

### PR DESCRIPTION
**Pull request by ZRazor**
I am targeting this branch, because it's BC.

Closes #357

## Changelog

```markdown
### Fixed
- Replaced FQCN strings with `::class` constants
- Used `createMock` in tests where it was possible
- Improved compatibility with SF 3.4, SF 4
```

### CODE REVIEW

**Comment by ZRazor** [comment]
Changed commit message
Fixed all stuff with `TwigRenderer` and  `FormRenderer`

**Comment by ZRazor** [comment]
@core23 Can you review this one, please

File: src/Controller/CategoryAdminController.php [hunk 1]
```diff
@@ -131,12 +136,19 @@ private function setFormTheme(FormView $formView, $theme)
...
+        if (!method_exists(AppVariable::class, 'getToken')) {
+            $twig->getExtension(FormExtension::class)
                 ->renderer->setTheme($formView, $theme);
 
             return;
         }
-        $twig->getRuntime('Symfony\Bridge\Twig\Form\TwigRenderer')->setTheme($formView, $theme);
+
+        // BC for Symfony < 3.4 where runtime should be TwigRenderer
+        if (!method_exists(DebugCommand::class, 'getLoaderPaths')) {
```
Comment (jordisala1991) [inline src/Controller/CategoryAdminController.php]: This check is not checking the same method right? can you check?

Comment (jordisala1991) [reply src/Controller/CategoryAdminController.php]: I checked and it is correct. so my fault

**Comment by jordisala1991** [comment]
Can you check why dev-master build is failing?

**Comment by ZRazor** [comment]
@jordisala1991 https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php#L220 
https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php#L281
it's now strict string type, and there is no stub for `getTemplate('list')`, so it fails with null.
I'll add it.

**Comment by ZRazor** [comment]
@jordisala1991 it's green now.

File: src/Admin/CollectionAdmin.php [hunk 2]
```diff
@@ -62,7 +62,7 @@ protected function configureFormFields(FormMapper $formMapper)
             ])
         ;
 
-        if (interface_exists('Sonata\MediaBundle\Model\MediaInterface')) {
+        if (interface_exists(MediaInterface::class)) {
```
Comment (core23) [inline src/Admin/CollectionAdmin.php]: Import is missing

File: src/Controller/Api/CategoryController.php [hunk 3]
```diff
@@ -256,7 +256,7 @@ protected function handleWriteCategory($request, $id = null)
 
             $view = FOSRestView::create($category);
 
-            if (class_exists('FOS\RestBundle\Context\Context')) {
+            if (class_exists(Context::class)) {
```
Comment (core23) [inline src/Controller/Api/CategoryController.php]: Import is missing

File: src/Controller/Api/CollectionController.php [hunk 4]
```diff
@@ -255,7 +255,7 @@ protected function handleWriteCollection($request, $id = null)
 
             $view = FOSRestView::create($collection);
 
-            if (class_exists('FOS\RestBundle\Context\Context')) {
+            if (class_exists(Context::class)) {
```
Comment (core23) [inline src/Controller/Api/CollectionController.php]: Import is missing

File: src/Controller/Api/ContextController.php [hunk 5]
```diff
@@ -255,7 +255,7 @@ protected function handleWriteContext($request, $id = null)
 
             $view = FOSRestView::create($context);
 
-            if (class_exists('FOS\RestBundle\Context\Context')) {
+            if (class_exists(Context::class)) {
```
Comment (core23) [inline src/Controller/Api/ContextController.php]: Import is missing

File: src/Controller/Api/CollectionController.php (hunk 4 above)
Comment (ZRazor) [reply src/Controller/Api/CollectionController.php]: It was already there https://github.com/sonata-project/SonataClassificationBundle/blob/3.x/src/Controller/Api/CollectionController.php#L14

File: src/Controller/Api/CategoryController.php (hunk 3 above)
Comment (ZRazor) [reply src/Controller/Api/CategoryController.php]: It was already there https://github.com/sonata-project/SonataClassificationBundle/blob/3.x/src/Controller/Api/CategoryController.php#L14

File: src/Admin/CollectionAdmin.php (hunk 2 above)
Comment (ZRazor) [reply src/Admin/CollectionAdmin.php]: I always check with code inspection for undefined classes, but yea - missed this one, thanks!

File: src/Controller/Api/ContextController.php (hunk 5 above)
Comment (ZRazor) [reply src/Controller/Api/ContextController.php]: It was already there https://github.com/sonata-project/SonataClassificationBundle/blob/3.x/src/Controller/Api/ContextController.php#L14

**Comment by OskarStark** [comment]
@greg0ire if pedantic, the changelog should be removed

**Comment by ZRazor** [comment]
@OskarStark I changed PR after @greg0ire added this label, so it's not pedantic now.

**Comment by Soullivaneuh** [comment]
@greg0ire It is pedantic. This is just code style changes and tests fixes.

**Comment by Soullivaneuh** [comment]
@greg0ire Well this is not. But this should not be merged like that @jordisala1991. In this case, this PR should have 3 commits. :wink:

**Comment by jordisala1991** [comment]
Indeed, we should’ve asked @ZRazor to split commits

**Pull request merged**